CommonClient.py

```diff
@@ -173,10 +174,69 @@ class CommonContext:
     items_handling: typing.Optional[int] = None
     want_slot_data: bool = True  # should slot_data be retrieved via Connect
 
-    # data package
-    # Contents in flux until connection to server is made, to download correct data for this multiworld.
-    item_names: typing.Dict[int, str] = Utils.KeyedDefaultDict(lambda code: f'Unknown item (ID:{code})')
-    location_names: typing.Dict[int, str] = Utils.KeyedDefaultDict(lambda code: f'Unknown location (ID:{code})')
+    class NameLookupDict:
+        """A specialized dict, with helper methods, for id -> name item/location data package lookups by game."""
+        _game_store: typing.Dict[str, typing.ChainMap[int, str]]
+
+        def __init__(self, ctx: CommonContext, lookup_type: typing.Literal["item", "location"]):
+            self.ctx = ctx
+            self.lookup_type = lookup_type
+            self._unknown_item = lambda key: f"Unknown {lookup_type} (ID: {key})"
+            self._archipelago_lookup = {}
+            self._game_store = collections.defaultdict(
+                lambda: collections.ChainMap(self._archipelago_lookup, Utils.KeyedDefaultDict(self._unknown_item)))
+            self._flat_store = Utils.KeyedDefaultDict(self._unknown_item)
+
+        # noinspection PyTypeChecker
+        def __getitem__(self, key: str) -> typing.Union[typing.ChainMap[int, str], typing.Dict[int, str]]:
```

Comment:
```suggestion
        def __getitem__(self, key: str) -> typing.Mapping[int, str]:
```
The return type was not wrong, but do we want to give that information about the internals in this public interface?
And I can't think of any good reason to mutate what is returned by this function.